File: without-library/graph/dag/dp_on_dag.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> adj[200005];
int memo[200005];

int longest_path(int u) {
    if(memo[u] != -1) return memo[u];
    
    int mx = 0;
    for(size_t i=0; i<adj[u].size(); i++) {
        int v = adj[u][i];
        int val = 1 + longest_path(v);
        if(val > mx) mx = val;
    }
    
    return memo[u] = mx;
}

int main() {
    return 0;
}

File: without-library/graph/dag/dp_on_dag.py
import sys
sys.setrecursionlimit(200000)

class DPDAG:
    def __init__(self, n, adj):
        self.n = n
        self.adj = adj
        self.memo = [-1] * (n + 1)
        
    def longest_path(self, u):
        if self.memo[u] != -1:
            return self.memo[u]
            
        mx = 0
        for v in self.adj[u]:
            val = 1 + self.longest_path(v)
            if val > mx: mx = val
            
        self.memo[u] = mx
        return mx

if __name__ == "__main__":
    pass
